docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+   }
+
+   // New
+   oe_select_attester_format(
+       requested_formats,
+       requested_formats_length,
+       &selected_format);
+```
+
+3. The application gets optional evidence parameters from verifiers and stays TEE-agnostic when generating evidences.
+
+```C
+   // Old
+   uint8_t opt_paramsX[] = { OPT_PARAMS_X };
+   uint8_t opt_paramsY[] = { OPT_PARAMS_Y };
+   if (format == ATTESTERX_UUID)
+   {
+      oe_get_evidence(&format, 0, NULL, 0, opt_paramsX, sizeof(opt_paramsX), &evidence, &evidence_size, NULL, NULL);
+   }
+   else if (format == ATTESTERY_UUID)
+   {
+       oe_get_evidence(&format, 0, NULL, 0, opt_paramsY, sizeof(opt_paramsY), &evidence, &evidence_size, NULL, NULL);
+   }
+
+   // New
+   uint8_t* opt_params;
+   size_t opt_params_size;
+   oe_get_evidence_parameters(&format, &opt_params, &opt_params_size);
+
+   oe_get_evidence(&format, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+```
+
+4. SGX report (referred as local attestation in Open Enclave) and SGX quote (referred as remote attestation in Open Enclave) generation takes different inputs and generates data of different structures. The local/remote attestation concept is SGX specific. To prevent confusion at evidence generation, the OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION and OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION flags were deprecated.
+
+```C
+   // Old
+   oe_get_evidence(&format_sgx, OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_sgx, OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_local_only, OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL); // Error
+   oe_get_evidence(&format_remote_only, OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL); // Error
+
+   // New
+   oe_get_evidence(&format_sgx_local, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_sgx_remote, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_local_only, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_remote_only, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+```
+
+User Experience
+---------------
+The user experience is improved by the changes. These are a couple of possible attestation scenarios:
+
+Scenario 1 - Attester selects from a list of evidence formats that can be verified:
+- Verifier application
+    1. Upon initialization, queries available verifiers to the application (oe_sgx_get_verifiers) and registers them (oe_register_verifier).
+    2. Builds the format IDs of the registered verifiers sorted by preference.
+    3. Sends the format IDs to the attester application. 
```

Comment:
I believe step 3 is not a good design.  There's no such concept currently in draft-ietf-rats-architecture.  If you believe there should be, then please bring up the rationale on the public rats@ietf.org mailing list.

On that topic, I would ask: how does the attester application get the URI (or IP/port or whatever) of the verifier application?   Why can't it get the knowledge the same way?   Also, we have to allow for the case where an attester is constrained and can only fit one format, so the verifier in general needs to support all formats so there's no need for an attester to learn in that case.

Reply:
This is one of the possible scenario for verifiers the can handle multiple evidence formats. For example, a server running a verifier that can verify SGX and TrustZone evidences, any client running in SGX or TrustZone enclaves can connect to that server and be attested.

The attester application can certainly get the evidence formats of the verifier from some sources. This is one of the possible scenario. It can be simplified with assumptions or complicated with more unknowns.

---
In reply to: [393252985](https://github.com/openenclave/openenclave/pull/2621#discussion_r393252985) [](ancestors = 393252985)

Reply:
I would encourage discussion on the IETF rats list if anyone believes this is actually an important scenario.

Reply:
I rephrased the user experience sentences to state that it's an example scenario.

One real life scenario that verifier tells attester what's supported:
- Bank teller (verifier) needs to check IDs to cash a personal check. The bank accepts driver's license or passport (2 formats).
- The customer (attester) reaches into his pocket and finds driver's license.

---
In reply to: [394652752](https://github.com/openenclave/openenclave/pull/2621#discussion_r394652752) [](ancestors = 394652752)

Reply:
Problem with that scenario is that the customer makes the decision of what to bring in the back long before talking to the teller (verifier).  That is, you either brought your passport or you didn't.   Once you get there, you pull out the one you have and use it.   At the time you signed up to be a bank customer, you agreed to the rules and you already know.  So you just present one thing because you know what the verifier needs.   As such, I don't yet see this as a real scenario.

Reply:
The scenarios shown here is just to explanation how the proposed APIs could be used. They arr not requirements. Its up to an application to design their own communication protocol between attesters and verifiers.

Reply:
The point is that the examples chosen don't seem to map to actual protocol recommendations, either in real-world scenarios or in IETF recommendations.  Instead, I would like to see examples more clearly map to actual protocol recommendations, in order to validate that the APIs would actually meet the requirements.